dbms/src/Core/Spiller.cpp

```diff
@@ -201,67 +202,84 @@ BlockInputStreams Spiller::restoreBlocks(UInt64 partition_id, UInt64 max_stream_
 {
     RUNTIME_CHECK_MSG(partition_id < partition_num, "{}: partition id {} exceeds partition num {}.", config.spill_id, partition_id, partition_num);
     RUNTIME_CHECK_MSG(isSpillFinished(), "{}: restore before the spiller is finished.", config.spill_id);
-    std::lock_guard partition_lock(spilled_files[partition_id]->spilled_files_mutex);
-    RUNTIME_CHECK_MSG(spilled_files[partition_id]->mutable_spilled_files.empty(), "{}: the mutable spilled files must be empty when restore.", config.spill_id);
-    auto & partition_spilled_files = spilled_files[partition_id]->immutable_spilled_files;
-
-    if (max_stream_size == 0)
-        max_stream_size = partition_spilled_files.size();
-    if (is_input_sorted && partition_spilled_files.size() > max_stream_size)
-    {
-        LOG_WARNING(logger, "Sorted spilled data restore does not take max_stream_size into account");
-    }
 
-    SpillDetails details{0, 0, 0};
     BlockInputStreams ret;
-    UInt64 spill_file_read_stream_num = is_input_sorted ? partition_spilled_files.size() : std::min(max_stream_size, partition_spilled_files.size());
-    std::vector<UInt64> restore_stream_read_rows;
-
-    if (is_input_sorted)
+    if unlikely (isAllConstant())
     {
-        for (auto & file : partition_spilled_files)
+        std::lock_guard lock(all_constant_mutex);
+        if (all_constant_block_rows[partition_id] > 0)
         {
-            RUNTIME_CHECK_MSG(file->exists(), "Spill file {} does not exists", file->path());
-            details.merge(file->getSpillDetails());
-            std::vector<SpilledFileInfo> file_infos;
-            file_infos.emplace_back(file->path());
-            restore_stream_read_rows.push_back(file->getSpillDetails().rows);
-            if (release_spilled_file_on_restore)
-                file_infos.back().file = std::move(file);
-            ret.push_back(std::make_shared<SpilledFilesInputStream>(std::move(file_infos), input_schema, header_without_constants, const_column_indexes, config.file_provider, spill_version));
+            // TODO use settings.max_block_size instead of DEFAULT_BLOCK_SIZE.
+            ret.push_back(std::make_shared<ConstantsBlockInputStream>(input_schema, all_constant_block_rows[partition_id], DEFAULT_BLOCK_SIZE));
+            all_constant_block_rows[partition_id] = 0;
```

Comment:
only set `all_constant_block_rows[partition_id]` to 0 if `release_spilled_file_on_restore ==true`?

Reply:
ok, updated.